src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/DefaultObjectValidator.cs

```diff
@@ -27,324 +21,49 @@ public class DefaultObjectValidator : IObjectModelValidator
-                foreach (var property in validationNode.ModelMetadata.Properties)
-                {
-                    var propertyExplorer = modelExplorer.GetExplorerForProperty(property.PropertyName);
-                    var propertyBindingName = property.BinderModelName ?? property.PropertyName;
-                    var childKey = ModelNames.CreatePropertyModelName(validationNode.Key, propertyBindingName);
-                    var childNode = new ModelValidationNode(childKey, property, propertyExplorer.Model)
-                    {
-                        ValidateAllProperties = true
-                    };
-                    childNodes.Add(childNode);
-                }
+                throw new ArgumentNullException(nameof(modelState));
             }
-            else
-            {
-                var enumerableModel = (IEnumerable)modelExplorer.Model;
-
-                // An integer index is incorrect in scenarios where there is a custom index provided by the user.
-                // However those scenarios are supported by createing a ModelValidationNode with the right keys.
-                var index = 0;
-                foreach (var element in enumerableModel)
-                {
-                    var elementExplorer = new ModelExplorer(_modelMetadataProvider, elementMetadata, element);
-                    var elementKey = ModelNames.CreateIndexModelName(validationNode.Key, index);
-                    var childNode = new ModelValidationNode(elementKey, elementMetadata, elementExplorer.Model)
-                    {
-                        ValidateAllProperties = true
-                    };
 
-                    childNodes.Add(childNode);
-                    index++;
-                }
-            }
+            var visitor = new ValidationVisitor(
+                validatorProvider,
+                _excludeFilters,
+                modelState,
+                validationState);
 
-            return childNodes;
-        }
-
-        private class ValidationContext
-        {
-            public ModelValidationContext ModelValidationContext { get; set; }
-
-            public HashSet<object> Visited { get; set; }
-
-            public ModelValidationNode ValidationNode { get; set; }
-        }
-
-        // Sorts validators based on whether or not they are 'required'. We want to run
-        // 'required' validators first so that we get the best possible error message.
-        private class ValidatorOrderComparer : IComparer<IModelValidator>
-        {
-            public static readonly ValidatorOrderComparer Instance = new ValidatorOrderComparer();
-
-            public int Compare(IModelValidator x, IModelValidator y)
-            {
-                var xScore = x.IsRequired ? 0 : 1;
-                var yScore = y.IsRequired ? 0 : 1;
-                return xScore.CompareTo(yScore);
-            }
+            var metadata = model == null ? null : _modelMetadataProvider.GetMetadataForType(model.GetType());
```

Comment:
Shouldn't this get the `ModelMetadata` from the `ValidationStateDictionary`, when that's available?